net: lib: coap: make ACK random factor runtime configurable

koalatux opened the pull request:
Extend the `coap_transmission_parameters` struct with the field `ack_random_percent`. This was the last remaining CoAP transmission parameter that was not configurable at runtime.

koalatux committed:
net: lib: coap: make ACK random factor runtime configurable

Extend the `coap_transmission_parameters` struct with the field
`ack_random_percent`. This was the last remaining CoAP transmission
parameter that was not configurable at runtime.

Signed-off-by: Adrian Friedli <adrian.friedli@husqvarnagroup.com>

rlubos commented:
I'm not sure if the last push was needed, those clang-format warning are not blockers, and IMHO the original formatting was more readable.

koalatux commented:
> I'm not sure if the last push was needed, those clang-format warning are not blockers, and IMHO the original formatting was more readable.

OK, I pushed the previous change again :-)

rlubos commented:
Good point, I think we could fallback to a default value if the runtime provided one is not considered correct (<= 100). Perhaps we should consider some upper bound too, it doesn't seem that spec defines it, but I guess we could set some arbitrary reasonable value (2? 5? I doubt any larger would make sense).

As for documentation, I guess an entry in the [migration guide](https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/migration-guide-4.0.rst) would be nice.

koalatux commented:
I added a fallback for the `ack_random_percent` field. If it is zero it defaults to the kconfig option `CONFIG_COAP_ACK_RANDOM_PERCENT`. Which is probably good enough, because when the struct is initialized using a designated initializer missing members get initialized with zero.

I also added a note in the migration guide.

pdgendt reviewed (dismissed):
Nit: the migration change should be in a separate commit.

koalatux committed:
doc: releases: migration-guide-4.0: add note about changed CoAP struct

Inform users about the new field in the `coap_transmission_parameters`
struct.

Signed-off-by: Adrian Friedli <adrian.friedli@husqvarnagroup.com>

jukkar commented on subsys/net/lib/coap/coap.c:
```diff
@@ -1706,8 +1709,9 @@ struct coap_pending *coap_pending_next_to_expire(
-				 CONFIG_COAP_ACK_RANDOM_PERCENT / 100;
+	const uint16_t random_percent = params->ack_random_percent ? params->ack_random_percent
+								   : CONFIG_COAP_ACK_RANDOM_PERCENT;
+	const uint32_t max_ack = params->ack_timeout * random_percent / 100;
```
Please change 100 to 100U so we do not need to change it later because of some tool complaining.
```
const uint32_t max_ack = params->ack_timeout * random_percent / 100U;
```
